src/cargo/ops/cargo_new.rs

```diff
@@ -67,22 +64,144 @@ pub fn new(opts: NewOptions, config: &Config) -> CargoResult<()> {
                         new_name);
                     try!(config.shell().say(&message, BLACK));
                 }
-                new_name
+                Ok(new_name)
             }
         }
-    };
+    }
+}
+
+fn check_name(name: &str) -> CargoResult<()> {
     for c in name.chars() {
         if c.is_alphanumeric() { continue }
         if c == '_' || c == '-' { continue }
         return Err(human(&format!("Invalid character `{}` in crate name: `{}`",
                                   c, name)));
     }
+    Ok(())
+}
+
+fn detect_source_path_and_type<'a : 'b, 'b>(project_path : &Path, 
+                                            project_name: &str, 
+                                            opts2: &'b mut NewOptions) -> CargoResult<()> {
```

Comment:
This argument should probably be called `opts` rather than `opts2`, and similarly if the two other arguments are renamed down below they can probably just have those names for the arguments